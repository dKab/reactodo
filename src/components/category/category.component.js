import React from 'react';
import Collapse from 'react-icons/fa/angle-down';
import Expand from 'react-icons/fa/angle-right';
import Edit from 'react-icons/fa/edit';
import Remove from 'react-icons/fa/trash';
import Plus from 'react-icons/fa/plus-square-o';
import Move from 'react-icons/fa/mail-reply';
import './category.css';
import Portal from 'react-modal';
import {DeleteCategoryModal} from '../delete-category-modal/delete-category-modal.component';
import {AddCategoryModal} from '../add-category-modal/add-category-modal.component';
import {ChangeCategoryNameModal} from '../change-category-name-modal/change-category-name-modal.component';
import {modalStyles} from '../modal/modal.component.jsx';

export const LIST_MODE = 'LIST_MODE';
export const DETAIL_MODE = 'DETAIL_MODE';
const ADD_MODAL = 'ADD_MODAL';
const DELETE_MODAL = 'DELETE_MODAL';
const EDIT_MODAL = 'EDIT_MODAL';

export class Category extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            [ADD_MODAL]: false,
            [ADD_MODAL]: false,
            [EDIT_MODAL]: false
        };
    }

    showModal(e, modal) {
        this.setState({...this.state, [modal]: true});
        e.stopPropagation();
    }

    confirmAddAndHideModal(name) {
        this.props.onPlusClick(name, this.props.category.id);
        this.hideModal(ADD_MODAL);
    }

    confirmDeleteAndHideModal() {
        this.props.onTrashClick(this.props.category.id);
        this.hideModal(DELETE_MODAL);
    }

    confirmEditAndHideModal(name) {
        this.props.onEditClick(this.props.category.id, name);
        this.hideModal(EDIT_MODAL);
    }

    hideModal(modal) {
        this.setState({...this.state, [modal]: false});
    }

    render() {
        let expandButton = null,
            isLeaf = this.props.isLeaf;
        if (this.props.category.expanded) {
            expandButton = <span className="category__expand-btn category__expand-btn--expanded"
                                 onClick={(e) => this.props.onExpandClick(this.props.category.id, e)}><Collapse /></span>;
        } else if (!isLeaf) {
            expandButton = <span className="category__expand-btn"
                                 onClick={(e) => this.props.onExpandClick(this.props.category.id, e)}><Expand /></span>
        }
        const classes = ['category'];
        if (isLeaf && this.props.category.parentId !== null) {
            classes.push('category--leaf');
        }
        if (this.props.selected) {
            classes.push('category--selected');
        }

        return (
            <div className={classes.join(' ')} onClick={() => this.props.onCategoryClick(this.props.category.id)}>
                {expandButton}
                <span className="category__name">{this.props.category.name}</span>
                <span className="category__edit-btn btn" onClick={(e) => this.showModal(e, EDIT_MODAL)}><Edit /></span>
                <div className="fr">
                    {this.props.mode === LIST_MODE &&
                    (<div><span className="category__remove-btn btn"
                                onClick={(e) => this.showModal(e, DELETE_MODAL)}><Remove /></span>
                        <span className="category__add-btn btn"
                              onClick={(e) => this.showModal(e, ADD_MODAL)}><Plus /></span></div>)}
                    {this.props.mode === DETAIL_MODE && <button className="category__move-btn btn"><Move /></button>}
                </div>
                <Portal style={modalStyles} contentLabel="modal" isOpen={this.state[ADD_MODAL]}>
                    <AddCategoryModal onCancel={() => this.hideModal(ADD_MODAL)} onConfirm={this.confirmAddAndHideModal.bind(this)} />
                </Portal>
                <Portal style={modalStyles} contentLabel="modal" isOpen={this.state[DELETE_MODAL]}>
                    <DeleteCategoryModal onCancel={() => this.hideModal(DELETE_MODAL)}
                         onConfirm={this.confirmDeleteAndHideModal.bind(this)}
                         categoryName={this.props.category.name} />
                </Portal>
                <Portal style={modalStyles} contentLabel="modal" isOpen={this.state[EDIT_MODAL]}>
                    <ChangeCategoryNameModal  onCancel={()=>this.hideModal(EDIT_MODAL)}
                        onConfirm={this.confirmEditAndHideModal.bind(this)}
                    />
                </Portal>
            </div>
        );
    }
}