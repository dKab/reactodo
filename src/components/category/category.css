.category {
    overflow: hidden;
}
.category--leaf .category__name {
    padding-left: 6px;
    display: inline-block;
}