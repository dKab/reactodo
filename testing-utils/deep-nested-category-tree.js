export const categories = [{
	"name": "cat1",
	"parentId": null,
	"id": 0,
	"selected": true,
	"expanded": true
}, {"name": "cat2", "parentId": null, "id": 1, "selected": false, "expanded": true}, {
	"name": "cat1_1",
	"parentId": 0,
	"id": 2,
	"selected": false,
	"expanded": false
}, {"name": "cat2_1", "parentId": 1, "id": 3, "selected": false, "expanded": false}, {
	"name": "cat2_2",
	"parentId": 1,
	"id": 4,
	"selected": false,
	"expanded": true
}, {"name": "cat2_2_1", "parentId": 4, "id": 5, "selected": false, "expanded": false}, {
	"name": "cat2_2_2",
	"parentId": 4,
	"id": 6,
	"selected": false,
	"expanded": true
}, {"name": "cat3", "parentId": null, "id": 7, "selected": false, "expanded": false}, {
	"name": "cat2_2_3",
	"parentId": 4,
	"id": 8,
	"selected": false,
	"expanded": false
}, {"name": "cat2_2_2_1", "parentId": 6, "id": 9, "selected": false, "expanded": true}, {
	"name": "cat2_2_2_2",
	"parentId": 6,
	"id": 10,
	"selected": false,
	"expanded": false
}, {"name": "cat2_2_2_1_1", "parentId": 9, "id": 11, "selected": false, "expanded": true}, {
	"name": "cat2_2_2_1_1_1",
	"parentId": 11,
	"id": 12,
	"selected": false,
	"expanded": false
}];