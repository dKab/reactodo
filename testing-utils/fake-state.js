export const fortyPercentsDoneState = {
    "todos": [
        {
            "name": "1",
            "description": "",
            "done": false,
            "categoryId": 0,
            "id": 0
        },
        {
            "name": "2",
            "description": "",
            "done": false,
            "categoryId": 0,
            "id": 1
        },
        {
            "name": "3",
            "description": "",
            "done": false,
            "categoryId": 0,
            "id": 2
        },
        {
            "name": "4",
            "description": "",
            "done": false,
            "categoryId": 0,
            "id": 3
        },
        {
            "name": "5",
            "description": "",
            "done": false,
            "categoryId": 0,
            "id": 4
        },
        {
            "name": "a",
            "description": "",
            "done": false,
            "categoryId": 1,
            "id": 5
        },
        {
            "name": "b",
            "description": "",
            "done": false,
            "categoryId": 1,
            "id": 6
        },
        {
            "name": "c",
            "description": "",
            "done": true,
            "categoryId": 2,
            "id": 7
        },
        {
            "name": "d",
            "description": "",
            "done": true,
            "categoryId": 3,
            "id": 8
        }
    ],
    "categories": [
        {
            "name": "cat1",
            "parentId": null,
            "id": 0,
            "selected": true,
            "expanded": false
        },
        {
            "name": "cat2",
            "parentId": null,
            "id": 1,
            "selected": false,
            "expanded": false
        },
        {
            "name": "cat3",
            "parentId": null,
            "id": 2,
            "selected": false,
            "expanded": false
        },
        {
            "name": "cat4",
            "parentId": null,
            "id": 3,
            "selected": false,
            "expanded": false
        },
        {
            "name": "cat5",
            "parentId": null,
            "id": 4,
            "selected": false,
            "expanded": false
        }
    ]
};

export const fortyPercentsDoneState2 = {
    "todos": [
        {
            "name": "1",
            "description": "",
            "done": false,
            "categoryId": 0,
            "id": 0
        },
        {
            "name": "2",
            "description": "",
            "done": false,
            "categoryId": 0,
            "id": 1
        },
        {
            "name": "3",
            "description": "",
            "done": true,
            "categoryId": 0,
            "id": 2
        },
        {
            "name": "4",
            "description": "",
            "done": true,
            "categoryId": 0,
            "id": 3
        },
        {
            "name": "5",
            "description": "",
            "done": false,
            "categoryId": 0,
            "id": 4
        },
        {
            "name": "a",
            "description": "",
            "done": false,
            "categoryId": 1,
            "id": 5
        },
        {
            "name": "b",
            "description": "",
            "done": true,
            "categoryId": 1,
            "id": 6
        },
        {
            "name": "c",
            "description": "",
            "done": true,
            "categoryId": 2,
            "id": 7
        },
        {
            "name": "d",
            "description": "",
            "done": true,
            "categoryId": 3,
            "id": 8
        }
    ],
    "categories": [
        {
            "name": "cat1",
            "parentId": null,
            "id": 0,
            "selected": true,
            "expanded": false
        },
        {
            "name": "cat2",
            "parentId": null,
            "id": 1,
            "selected": false,
            "expanded": false
        },
        {
            "name": "cat3",
            "parentId": null,
            "id": 2,
            "selected": false,
            "expanded": false
        },
        {
            "name": "cat4",
            "parentId": null,
            "id": 3,
            "selected": false,
            "expanded": false
        },
        {
            "name": "cat5",
            "parentId": null,
            "id": 4,
            "selected": false,
            "expanded": false
        }
    ]
};

export const eightyPercentsDoneState = {
    "todos": [
        {
            "name": "1",
            "description": "",
            "done": true,
            "categoryId": 0,
            "id": 0
        },
        {
            "name": "2",
            "description": "",
            "done": true,
            "categoryId": 0,
            "id": 1
        },
        {
            "name": "3",
            "description": "",
            "done": true,
            "categoryId": 0,
            "id": 2
        },
        {
            "name": "4",
            "description": "",
            "done": true,
            "categoryId": 0,
            "id": 3
        },
        {
            "name": "5",
            "description": "",
            "done": true,
            "categoryId": 0,
            "id": 4
        },
        {
            "name": "a",
            "description": "",
            "done": false,
            "categoryId": 1,
            "id": 5
        },
        {
            "name": "b",
            "description": "",
            "done": true,
            "categoryId": 1,
            "id": 6
        },
        {
            "name": "c",
            "description": "",
            "done": true,
            "categoryId": 2,
            "id": 7
        },
        {
            "name": "d",
            "description": "",
            "done": true,
            "categoryId": 3,
            "id": 8
        },
        {
            "name": "fdsaf",
            "description": "",
            "done": true,
            "categoryId": 4,
            "id": 9
        }
    ],
    "categories": [
        {
            "name": "cat1",
            "parentId": null,
            "id": 0,
            "selected": true,
            "expanded": false
        },
        {
            "name": "cat2",
            "parentId": null,
            "id": 1,
            "selected": false,
            "expanded": false
        },
        {
            "name": "cat3",
            "parentId": null,
            "id": 2,
            "selected": false,
            "expanded": false
        },
        {
            "name": "cat4",
            "parentId": null,
            "id": 3,
            "selected": false,
            "expanded": false
        },
        {
            "name": "cat5",
            "parentId": null,
            "id": 4,
            "selected": false,
            "expanded": false
        }
    ]
};