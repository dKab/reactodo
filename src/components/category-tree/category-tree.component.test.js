import renderer from 'react-test-render';
import CategoryTree from './category-tree.component';
import {categories} from '../../../testing-utils/deep-nested-category-tree';
import {LIST_MODE, DETAIL_MODE} from '../category/category.component'

test('CategogryTree renders correctly', () => {
	const tree = renderer.create(
		<CategoryTree categories={categories} categoriesMode={LIST_MODE} selectedCategory={'1'} todo={{}} />
	).toJSON();
	expect(tree).toMatchSnapshot();
});
