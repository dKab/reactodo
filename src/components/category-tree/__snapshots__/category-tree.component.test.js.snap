exports[`test CategogryTree renders correctly 1`] = `
<div
  className="category-tree">
  <div
    style={
      Object {
        "marginLeft": "0px",
      }
    }>
    <div
      className="category"
      onClick={[Function]}>
      <span
        className="category__expand-btn category__expand-btn--expanded"
        onClick={[Function]}>
        <span
          className="fa fa-angle-down" />
      </span>
      <span
        className="category__name">
        cat1
      </span>
      <span
        className="category__edit-btn btn"
        onClick={[Function]}>
        <span
          className="fa fa-edit" />
      </span>
      <div
        className="fr">
        <div>
          <span
            className="category__remove-btn btn"
            onClick={[Function]}>
            <span
              className="fa fa-trash" />
          </span>
          <span
            className="category__add-btn btn"
            onClick={[Function]}>
            <span
              className="fa fa-plus-square-o" />
          </span>
        </div>
      </div>
      <noscript />
      <noscript />
      <noscript />
    </div>
    <div
      style={
        Object {
          "marginLeft": "10px",
        }
      }>
      <div
        className="category category--leaf"
        onClick={[Function]}>
        <span
          className="category__name">
          cat1_1
        </span>
        <span
          className="category__edit-btn btn"
          onClick={[Function]}>
          <span
            className="fa fa-edit" />
        </span>
        <div
          className="fr">
          <div>
            <span
              className="category__remove-btn btn"
              onClick={[Function]}>
              <span
                className="fa fa-trash" />
            </span>
            <span
              className="category__add-btn btn"
              onClick={[Function]}>
              <span
                className="fa fa-plus-square-o" />
            </span>
          </div>
        </div>
        <noscript />
        <noscript />
        <noscript />
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "marginLeft": "0px",
      }
    }>
    <div
      className="category"
      onClick={[Function]}>
      <span
        className="category__expand-btn category__expand-btn--expanded"
        onClick={[Function]}>
        <span
          className="fa fa-angle-down" />
      </span>
      <span
        className="category__name">
        cat2
      </span>
      <span
        className="category__edit-btn btn"
        onClick={[Function]}>
        <span
          className="fa fa-edit" />
      </span>
      <div
        className="fr">
        <div>
          <span
            className="category__remove-btn btn"
            onClick={[Function]}>
            <span
              className="fa fa-trash" />
          </span>
          <span
            className="category__add-btn btn"
            onClick={[Function]}>
            <span
              className="fa fa-plus-square-o" />
          </span>
        </div>
      </div>
      <noscript />
      <noscript />
      <noscript />
    </div>
    <div
      style={
        Object {
          "marginLeft": "10px",
        }
      }>
      <div
        className="category category--leaf"
        onClick={[Function]}>
        <span
          className="category__name">
          cat2_1
        </span>
        <span
          className="category__edit-btn btn"
          onClick={[Function]}>
          <span
            className="fa fa-edit" />
        </span>
        <div
          className="fr">
          <div>
            <span
              className="category__remove-btn btn"
              onClick={[Function]}>
              <span
                className="fa fa-trash" />
            </span>
            <span
              className="category__add-btn btn"
              onClick={[Function]}>
              <span
                className="fa fa-plus-square-o" />
            </span>
          </div>
        </div>
        <noscript />
        <noscript />
        <noscript />
      </div>
    </div>
    <div
      style={
        Object {
          "marginLeft": "10px",
        }
      }>
      <div
        className="category"
        onClick={[Function]}>
        <span
          className="category__expand-btn category__expand-btn--expanded"
          onClick={[Function]}>
          <span
            className="fa fa-angle-down" />
        </span>
        <span
          className="category__name">
          cat2_2
        </span>
        <span
          className="category__edit-btn btn"
          onClick={[Function]}>
          <span
            className="fa fa-edit" />
        </span>
        <div
          className="fr">
          <div>
            <span
              className="category__remove-btn btn"
              onClick={[Function]}>
              <span
                className="fa fa-trash" />
            </span>
            <span
              className="category__add-btn btn"
              onClick={[Function]}>
              <span
                className="fa fa-plus-square-o" />
            </span>
          </div>
        </div>
        <noscript />
        <noscript />
        <noscript />
      </div>
      <div
        style={
          Object {
            "marginLeft": "10px",
          }
        }>
        <div
          className="category category--leaf"
          onClick={[Function]}>
          <span
            className="category__name">
            cat2_2_1
          </span>
          <span
            className="category__edit-btn btn"
            onClick={[Function]}>
            <span
              className="fa fa-edit" />
          </span>
          <div
            className="fr">
            <div>
              <span
                className="category__remove-btn btn"
                onClick={[Function]}>
                <span
                  className="fa fa-trash" />
              </span>
              <span
                className="category__add-btn btn"
                onClick={[Function]}>
                <span
                  className="fa fa-plus-square-o" />
              </span>
            </div>
          </div>
          <noscript />
          <noscript />
          <noscript />
        </div>
      </div>
      <div
        style={
          Object {
            "marginLeft": "10px",
          }
        }>
        <div
          className="category"
          onClick={[Function]}>
          <span
            className="category__expand-btn category__expand-btn--expanded"
            onClick={[Function]}>
            <span
              className="fa fa-angle-down" />
          </span>
          <span
            className="category__name">
            cat2_2_2
          </span>
          <span
            className="category__edit-btn btn"
            onClick={[Function]}>
            <span
              className="fa fa-edit" />
          </span>
          <div
            className="fr">
            <div>
              <span
                className="category__remove-btn btn"
                onClick={[Function]}>
                <span
                  className="fa fa-trash" />
              </span>
              <span
                className="category__add-btn btn"
                onClick={[Function]}>
                <span
                  className="fa fa-plus-square-o" />
              </span>
            </div>
          </div>
          <noscript />
          <noscript />
          <noscript />
        </div>
        <div
          style={
            Object {
              "marginLeft": "10px",
            }
          }>
          <div
            className="category"
            onClick={[Function]}>
            <span
              className="category__expand-btn category__expand-btn--expanded"
              onClick={[Function]}>
              <span
                className="fa fa-angle-down" />
            </span>
            <span
              className="category__name">
              cat2_2_2_1
            </span>
            <span
              className="category__edit-btn btn"
              onClick={[Function]}>
              <span
                className="fa fa-edit" />
            </span>
            <div
              className="fr">
              <div>
                <span
                  className="category__remove-btn btn"
                  onClick={[Function]}>
                  <span
                    className="fa fa-trash" />
                </span>
                <span
                  className="category__add-btn btn"
                  onClick={[Function]}>
                  <span
                    className="fa fa-plus-square-o" />
                </span>
              </div>
            </div>
            <noscript />
            <noscript />
            <noscript />
          </div>
          <div
            style={
              Object {
                "marginLeft": "10px",
              }
            }>
            <div
              className="category"
              onClick={[Function]}>
              <span
                className="category__expand-btn category__expand-btn--expanded"
                onClick={[Function]}>
                <span
                  className="fa fa-angle-down" />
              </span>
              <span
                className="category__name">
                cat2_2_2_1_1
              </span>
              <span
                className="category__edit-btn btn"
                onClick={[Function]}>
                <span
                  className="fa fa-edit" />
              </span>
              <div
                className="fr">
                <div>
                  <span
                    className="category__remove-btn btn"
                    onClick={[Function]}>
                    <span
                      className="fa fa-trash" />
                  </span>
                  <span
                    className="category__add-btn btn"
                    onClick={[Function]}>
                    <span
                      className="fa fa-plus-square-o" />
                  </span>
                </div>
              </div>
              <noscript />
              <noscript />
              <noscript />
            </div>
            <div
              style={
                Object {
                  "marginLeft": "10px",
                }
              }>
              <div
                className="category category--leaf"
                onClick={[Function]}>
                <span
                  className="category__name">
                  cat2_2_2_1_1_1
                </span>
                <span
                  className="category__edit-btn btn"
                  onClick={[Function]}>
                  <span
                    className="fa fa-edit" />
                </span>
                <div
                  className="fr">
                  <div>
                    <span
                      className="category__remove-btn btn"
                      onClick={[Function]}>
                      <span
                        className="fa fa-trash" />
                    </span>
                    <span
                      className="category__add-btn btn"
                      onClick={[Function]}>
                      <span
                        className="fa fa-plus-square-o" />
                    </span>
                  </div>
                </div>
                <noscript />
                <noscript />
                <noscript />
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "marginLeft": "10px",
            }
          }>
          <div
            className="category category--leaf"
            onClick={[Function]}>
            <span
              className="category__name">
              cat2_2_2_2
            </span>
            <span
              className="category__edit-btn btn"
              onClick={[Function]}>
              <span
                className="fa fa-edit" />
            </span>
            <div
              className="fr">
              <div>
                <span
                  className="category__remove-btn btn"
                  onClick={[Function]}>
                  <span
                    className="fa fa-trash" />
                </span>
                <span
                  className="category__add-btn btn"
                  onClick={[Function]}>
                  <span
                    className="fa fa-plus-square-o" />
                </span>
              </div>
            </div>
            <noscript />
            <noscript />
            <noscript />
          </div>
        </div>
      </div>
      <div
        style={
          Object {
            "marginLeft": "10px",
          }
        }>
        <div
          className="category category--leaf"
          onClick={[Function]}>
          <span
            className="category__name">
            cat2_2_3
          </span>
          <span
            className="category__edit-btn btn"
            onClick={[Function]}>
            <span
              className="fa fa-edit" />
          </span>
          <div
            className="fr">
            <div>
              <span
                className="category__remove-btn btn"
                onClick={[Function]}>
                <span
                  className="fa fa-trash" />
              </span>
              <span
                className="category__add-btn btn"
                onClick={[Function]}>
                <span
                  className="fa fa-plus-square-o" />
              </span>
            </div>
          </div>
          <noscript />
          <noscript />
          <noscript />
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "marginLeft": "0px",
      }
    }>
    <div
      className="category"
      onClick={[Function]}>
      <span
        className="category__name">
        cat3
      </span>
      <span
        className="category__edit-btn btn"
        onClick={[Function]}>
        <span
          className="fa fa-edit" />
      </span>
      <div
        className="fr">
        <div>
          <span
            className="category__remove-btn btn"
            onClick={[Function]}>
            <span
              className="fa fa-trash" />
          </span>
          <span
            className="category__add-btn btn"
            onClick={[Function]}>
            <span
              className="fa fa-plus-square-o" />
          </span>
        </div>
      </div>
      <noscript />
      <noscript />
      <noscript />
    </div>
  </div>
</div>
`;
